rfcs/20200710-reduction/README.md

```diff
@@ -0,0 +1,159 @@
+Proposal for Reduction primitive
+====================================================
+
+# Introduction
+
+Reduction functionality motivated by the following requests from OpenVino team:
+* Add support for Reduce operations:
+  - ReduceLogicalAnd
+  - ReduceLogicalOr
+  - ReduceMax
+  - ReduceMean
+  - ReduceMin
+  - ReduceProd
+  - ReduceSum
+* Add support for Normalize operations:
+  - L-2 normalization
+
+In addition to L-2 the following Normalize agorithms are nice-to-have:
+  - L-1 normalization
+  - L-infinity normalization
+
+The following Reduce and Norm algorithms are supported by the most popular
+frameworks:
+
+| Algotirhm | Formula                      | TensorFlow | PyTorch | OpenVino |
+| ---------:| ----------------------------:| ----------:| -------:| --------:|
+| And       | and(x[:])                    |            |         | x        |
+| Or        | or(x[:])                     |            |         | x        |
+| Min       | min(x[:])                    | x          | x       | x        |
+| Max       | max(x[:])                    | x          | x       | x        |
+| Sum       | sum(x[:])                    | x          | x       | x        |
+| Prod      | mul(x[:])                    | x          | x       | x        |
+| Mean      | sum(x[:]) / n                | x          | x       | x        |
+| Variance  | Mean(pow(x[:] - Mean(x), 2)) | x          | x       |          |
+| Std       | pow(Variance(x), 0.5)        | x          | x       |          |
+| LogSumExp | log(sum(exp(x[:] - Max(x)))) + Max(x) | x | x       |          |
+| Median    |                              |            | x       |          |
+| Mode      |                              |            | x       |          |
+| p-norm    | pow(sum(pow(abs(x[:]), p)), 1/p) | x      | x       |          |
+
+The following algorithms are out of scope due to low priority:
+* And;
+* Or;
+* Median;
+* Mode.
+
+Also there are normalize operations, which variate between frameworks:
+* OpenVino:
+  L2_normalize(x) = x[:] / pow(eps_mode(sum(pow(x[:], 2)), eps), 0.5)
+    = x[:] / pow(eps_mode((pow(norm_l2(x[:]), 2)), eps), 0.5),
```

Comment:
Is this mode covered by the option 1?

Reply:
Yes, it can be covered using post ops:
L2_normalize(x) = reduction(alg:norm_l2, x) + POST_OPS(eltwise(alg:pow, 2) + binary(alg:{max, add}) + eltwise(alg:pow, 0.5))

Reply:
Thanks!